psh/test-ls.py

```diff
@@ -0,0 +1,153 @@
+    psh.assert_cmd(p, cmd, '', msg)
+
+    msg = 'Prompt not seen after calling `mkdir test_ls_dir/dir`'
+    psh.assert_cmd(p, 'mkdir test_ls_dir/dir', r'.*?(\r+\n)?', msg, is_regex=True)
+
+
+def assert_ls_err(p):
+    msg = 'Wrong error message, when trying to print nonexistent directory content'
+    psh.assert_cmd(p, 'ls nonexistentDir', "ls: can't access nonexistentDir: no such file or directory", msg)
+
+    return True
+
+
+def create_pattern(files, separate_lines=False):
+    new_line_reg = r'(\r+\n)'
+    if not separate_lines:
+        new_line_reg += r'?'
+    pattern = r''
+    for file in files:
+        # 0 or 1 esc code + filename + white spaces or esc code (one or more time) + newline (optional or not)
+        pattern += r'(\x1b\[\S*)?' + f'{file}' + r'((\s*)|(\x1b\[\S*))+' + new_line_reg
+    if not separate_lines:
+        pattern += psh.EOL
+    return pattern
+
+
+def assert_ls_noarg(p, files):
+    expected = create_pattern(sorted(files))
+    msg = 'Wrong content of listed directory, when calling ls without arguments'
+    psh.assert_cmd(p, 'ls test_ls_dir', expected, msg, is_regex=True)
+
+
+def assert_ls_1(p, files):
+    expected = create_pattern(sorted(files), separate_lines=True)
+    msg = 'Wrong content of listed directory, when calling `ls -1`'
+    psh.assert_cmd(p, 'ls -1 test_ls_dir', expected, msg, is_regex=True)
+
+
+def assert_ls_a(p, files):
+    expected = create_pattern(sorted(files))
+    msg = 'Wrong content of listed directory, when calling `ls -a`'
+    psh.assert_cmd(p, 'ls -a test_ls_dir', expected, msg, is_regex=True)
+
+
+def assert_ls_d(p):
+    expected = r'(\x1b\[\S*)?' + 'test_ls_dir' + r'\s*' + r'(\x1b\[\S*)?' + r'(\r+\n)?'
+    msg = 'Wrong output, when calling `ls -d tested_directory`'
+    psh.assert_cmd(p, 'ls -d test_ls_dir', expected, msg, is_regex=True)
+
+
+def assert_ls_d_noarg(p):
+    expected = r'(\x1b\[\S*)?' + '.' + r'\s*' + r'(\x1b\[\S*)?' + r'(\r+\n)?'
+    msg = 'Wrong output, when calling `ls -d` without specified directory'
+    psh.assert_cmd(p, 'ls -d', expected, msg, is_regex=True)
+
+
+def assert_ls_f(p, files):
+    expected = create_pattern(files)
+    msg = 'Wrong output, when calling `ls -f test_ls_dir`'
+    psh.assert_cmd(p, 'ls -f test_ls_dir', expected, msg, is_regex=True)
+
+
+def assert_ls_h(p):
+    expected = r'usage:(\s)ls \[options\] \[files\](\r+\n)' + \
+                r'(\s*-[\w\s]:[ \S]+(\r+\n))+'
+    msg = 'Wrong ls help message format'
+    psh.assert_cmd(p, 'ls -h', expected, msg, is_regex=True)
+
+
+def assert_ls_l(p, files):
+    cmd = 'ls -l test_ls_dir'
+    p.sendline(cmd)
+    p.expect(cmd)
+    for file in sorted(files):
+        if file[0] == '.':
+            continue
+        p.expect(r'(-|b|c|d|l|n|p|s|D|E|O|S)([-r][-w][-x]){3}( )'
+                 + r'(\d)( )'
+                 + r'(root |--- ){2}( *)'
+                 + r'(\d+)( )'
+                 + r'(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)( )'
```

Comment:
There can be two spaces before date if listening contains files with two digit days